tangelo/linq/tests/test_simulator.py

```diff
@@ -636,6 +640,140 @@ def backend_info():
+            assert_freq_dict_almost_equal(f, {"11": 1}, 1.e-7)
+            assert_freq_dict_almost_equal(circuit.success_probabilities, {"001": 0.125}, 1.e-7)
+            self.assertTrue(applied_circuit == Circuit(circuit.applied_gates))
+
+            # Test that the correct ordering of probabilities is obtained when running without specifying the
+            # desired measurement result.
+            sim.n_shots = 1000
+
+            circuit = Circuit([Gate("H", 0), Gate("CMEASURE", 0),  Gate("CNOT", 1, 0)], cmeasure_control=cfunc)
+            f, _ = sim.simulate(circuit, save_mid_circuit_meas=True)
+            assert_freq_dict_almost_equal(f, {"11": 1}, 1.e-7)
+            self.assertTrue(sim.mid_circuit_meas_freqs["1"] > sim.mid_circuit_meas_freqs["01"])
+            self.assertTrue(sim.mid_circuit_meas_freqs["01"] > sim.mid_circuit_meas_freqs["001"])
+
+    def test_measurement_controlled_gates_class(self):
+        """Test of an adaptive IterativeQPE circuit."""
+
+        for backend in installed_cmeasure_simulators:
+
+            sim = get_backend(backend, n_shots=2)
+
+            # Unitary has eigenvalue (1/16+1/8+1/2)*pi with eigenvector |1> on qubit 1.
+            # Measurement will be 0011010 = 0/64 + 0/32 + 1/16 + 1/8 + 0/4 + 1/2 + 0
+            ugate = Gate("CPHASE", 1, control=0, parameter=np.pi*(1/16 + 1/8 + 1/2))
+
+            class IterativeQPE(ClassicalControl):
+                def __init__(self, n_bits: int):
+                    """Iterative QPE with n_bits"""
+                    self.n_bits: int = n_bits
+                    self.bitplace: int = copy(self.n_bits)
+                    self.phase: float = 0
+                    self.measurements: List[str] = [""]
+                    self.energies: List[float] = [0.]
+                    self.n_runs: int = 0
+
+                def return_circuit(self, measurement) -> List[Gate]:
+                    self.measurements[self.n_runs] += measurement
+                    self.energies[self.n_runs] += int(measurement)/2**self.bitplace
+
+                    if self.bitplace > 0:
+                        self.bitplace -= 1
+
+                        # Update phase and determine reset gates
+                        if measurement == "1":
+                            self.phase += 1/2**(self.bitplace+1)
+                            reset_to_zero = [Gate("X", 0)]
+                        else:
+                            reset_to_zero = []
+
+                        phase_correction = [Gate("PHASE", 0, parameter=-np.pi*self.phase*2**(self.bitplace))]
+                        gates = reset_to_zero + [Gate("H", 0)] + phase_correction + [ugate]*2**self.bitplace + [Gate("H", 0)]
+                        return gates + [Gate("CMEASURE", 0)] if self.bitplace else gates
+                    else:
+                        return []
+
+                def finalize(self):
+                    self.bitplace = copy(self.n_bits)
```

Comment:
no need for copy. self.n_bits is immutable so you are getting a new variable pointing to that specific value.